### PR TITLE
Die error_chain, die

### LINKED ISSUE #10302: Replace `error-chain`
`error-chain` is not maintained anymore and depends on [Error::cause](https://doc.rust-lang.org/std/error/trait.Error.html#method.cause) which will be deprecated in Rust 1.33.

I don't know what the best solution to this but what I can think of:
- Manually implement it
- Migrate to `failure`

### PR DESCRIPTION
**Pull request by ascjones**
Resolves #10302.

### CODE REVIEW

File: util/network/src/error.rs [hunk 1]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	#[display(fmt ="Packet is too large")]
+	OversizedPacket,
+	/// Reached system resource limits for this process
+	#[display(fmt ="Too many open files in this process. Check your resource limits and restart parity")]
+	ProcessTooManyFiles,
+	/// Reached system wide resource limits
+	#[display(fmt ="Too many open files on system. Consider closing some processes/release some file handlers or increas the system-wide resource limits and restart parity.")]
+	SystemTooManyFiles,
+	/// An unknown IO error occurred.
+	#[display(fmt ="Unexpected IO error: {}", _0)]
```
Comment (ordian) [inline util/network/src/error.rs]: There is a space missing after `=` here and above.

File: util/network/src/error.rs [hunk 2]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+#[derive(Debug, derive_more::Display)]
+pub enum Error {
+	/// Socket IO error.
+	SocketIo(IoError),
+	/// Decompression error.
+	Decompression(snappy::InvalidInput),
+	/// Rlp decoder error.
+	Rlp(rlp::DecoderError),
+	/// Error concerning the network address parsing subsystem.
+	#[display(fmt ="Failed to parse network address")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Failed to parse network address")]
```

File: util/network/src/error.rs [hunk 3]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	AddressParse,
+	/// Error concerning the network address resolution subsystem.
+	#[display(fmt ="Failed to resolve network address {}", _0)]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Failed to resolve network address {}", _0)]
```

File: util/network/src/error.rs [hunk 4]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	AddressResolve(AddressResolveError),
+	/// Authentication failure
+	#[display(fmt ="Authentication failure")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Authentication failure")]
```

File: util/network/src/error.rs [hunk 5]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	Auth,
+	/// Unrecognised protocol
+	#[display(fmt ="Bad protocol")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Bad protocol")]
```

File: util/network/src/error.rs [hunk 6]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	BadProtocol,
+	/// Expired message
+	#[display(fmt ="Expired message")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Expired message")]
```

File: util/network/src/error.rs [hunk 7]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	Expired,
+	/// Peer not found
+	#[display(fmt ="Peer not found")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Peer not found")]
```

File: util/network/src/error.rs [hunk 8]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	PeerNotFound,
+	/// Peer is disconnected
+	#[display(fmt ="Peer disconnected: {}", _0)]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Peer disconnected: {}", _0)]
```

File: util/network/src/error.rs [hunk 9]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	Disconnect(DisconnectReason),
+	/// Invalid node id
+	#[display(fmt ="Invalid node id")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Invalid node id")]
```

File: util/network/src/error.rs [hunk 10]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	InvalidNodeId,
+	/// Packet size is over the protocol limit
+	#[display(fmt ="Packet is too large")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Packet is too large")]
```

File: util/network/src/error.rs [hunk 11]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	OversizedPacket,
+	/// Reached system resource limits for this process
+	#[display(fmt ="Too many open files in this process. Check your resource limits and restart parity")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Too many open files in this process. Check your resource limits and restart parity")]
```

File: util/network/src/error.rs [hunk 12]
```diff
@@ -85,118 +81,127 @@ impl fmt::Display for DisconnectReason {
...
+	ProcessTooManyFiles,
+	/// Reached system wide resource limits
+	#[display(fmt ="Too many open files on system. Consider closing some processes/release some file handlers or increas the system-wide resource limits and restart parity.")]
```
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Too many open files on system. Consider closing some processes/release some file handlers or increas the system-wide resource limits and restart parity.")]
```

File: util/network/src/error.rs (hunk 1 above)
Comment (dvdplm) [inline util/network/src/error.rs]:
```suggestion
	#[display(fmt = "Unexpected IO error: {}", _0)]
```

**Comment by ascjones** [comment]
@dvdplm just killed all those

**Pull request merged**